### PR TITLE
and fast-gru and fast-pgru layer

### PR DESCRIPTION
**Pull request by hlthu**

### CODE REVIEW

File: egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py [hunk 1]
```diff
@@ -411,4 +412,348 @@ def generate_pgru_config(self):
...
+                      " zeroing-interval={2}"
+                      " recurrence-interval={3}"
+                      "".format(self.config['clipping-threshold'],
+                                self.config['zeroing-threshold'],
+                                self.config['zeroing-interval'], abs(delay)))
+        repair_nonlin = self.config['self-repair-scale-nonlinearity']
+        repair_nonlin_str = "self-repair-scale={0:.10f}".format(repair_nonlin) if repair_nonlin is not None else ''
+        affine_str = self.config['ng-affine-options']
+
+        # string for GruNonlinearityComponent
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py:516]: For here, try the style used in Fast-LSTM, see https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/steps/libs/nnet3/xconfig/lstm.py#L541, avoiding bunch of if sentence

File: egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py [hunk 2]
```diff
@@ -411,4 +412,348 @@ def generate_pgru_config(self):
...
+        # h = \tanh ( x_t * U^h + s_{t-1} \dot r * W^h )
+        # c_t = ( 1 - z ) \dot h + z \dot y_{t-1}
+        # y_t = c_t * W^s
+        # s_t = y_t(0:rec_proj_dim-1)
+        
+        
+
+        configs = []
+        configs.append("# W_zr matrics for z_t and r_t")
+        configs.append("component name={0}.W_zr type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input_dim + rec_proj_dim, cell_dim + rec_proj_dim, affine_str))
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: You merge the two matrix into one, do you have to change the default 'max-change=0.75' ? increase it a little ?

File: egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py [hunk 3]
```diff
@@ -411,4 +412,348 @@ def generate_pgru_config(self):
...
+        configs.append("# hpart_t")
+        configs.append("component-node name={0}.hpart_t component={0}.W_hpart input={1}".format(name, input_descriptor))
+        
+        configs.append("# c_t and h_t")
+        configs.append("component name={0}.gru_nonlin type=GruNonlinearityComponent cell-dim={1} recurrent-dim={2} {3}".format(name, cell_dim, rec_proj_dim, gru_nonlin_str))
+        configs.append("component-node name={0}.gru_nonlin_t component={0}.gru_nonlin input=Append({0}.zr_t, {0}.hpart_t, IfDefined(Offset({0}.c_trunc, {2})), IfDefined(Offset({1}, {2})))".format(name, recurrent_connection, delay))
+        configs.append("dim-range-node name={0}.c_t input-node={0}.gru_nonlin_t dim-offset={1} dim={1}".format(name, cell_dim))
+
+        configs.append("# c_trunc")
+        configs.append("component name={0}.c_trunc type=BackpropTruncationComponent dim={1} {2}".format(name, cell_dim, bptrunc_str))
```
Comment (GaofengCheng) [inline egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: removing this trunc component

Comment (hlthu) [reply egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: Ok.

File: egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py (hunk 2 above)
Comment (hlthu) [reply egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: I will first change it to 1.0 and then do some experiments to determine the final value.

File: egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py (hunk 1 above)
Comment (hlthu) [reply egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py:516]: Ok,i will use gru-nonlinearity-options

**Comment by hlthu** [comment]
fast (p)gru Vs. no-fast (p)gru under TEDLIUM-v2

|       AM       | dev WER% | test WER% | #Parameters | Train TIme | Decode dev | Decode test |
| :------------: | :------: | :-------: | :---------: | :--------: | :--------: | :---------: |
|   bpgru-256    |   8.1    |    8.4    |    29.4M    |   30.7h    |    3.2h    |    5.1h     |
|    bgru-600    |   9.0    |    9.2    |    30.1M    |   28.8h    |    3.2h    |    5.2h     |
| fast-bpgru-256 |   8.4    |    8.8    |    29.4M    |   26.2h    |    2.4h    |    3.8h     |
| fast-bgru-600  |   8.9    |    9.4    |    30.1M    |   24.0h    |    2.6h    |    4.0h     |

note: 

- pgru-256 means pgru with 1024 cell and both 256 rec and non_rec dim; 
- gru-600 means gru with 600 cell;
- Train time is all the time of GPU computing, which is the sum of the time in train.\*.log;
- Decode time is also the sum time of all decode threads, which is the sum of the time in decode.\*.log, and there we use 38 threads.

conclusions:

* When trainning, the fast version can save about 17% time for gru, and 15% for pgru;
* When decoding, the fast version can save about 23% time for gru, and 25% for pgru.

File: egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py (hunk 2 above)
Comment (hlthu) [reply egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: @GaofengCheng Shall I need to change the default 'max-change=0.10', it seems that it doesn't work well for fast-pgru.

**Comment by hlthu** [comment]
And I use the `cat progress.10.log | grep diff`
for fast-bpgru:
LOG (nnet3-show-progress[5.2.6-e93e]:main():nnet3-show-progress.cc:139) Parameter differences per layer are [ blstm1-forward.W_zr:1.98216 blstm1-forward.W_hpart:1.53073 blstm1-forward.gru_nonlin:2.10125 blstm1-forward.W_y:2.95711 blstm1-backward.W_zr:0.702571 blstm1-backward.W_hpart:0.628286 blstm1-backward.gru_nonlin:1.27958 blstm1-backward.W_y:2.40449 blstm2-forward.W_zr:0.949411 blstm2-forward.W_hpart:2.00341 blstm2-forward.gru_nonlin:0.871313 blstm2-forward.W_y:2.00235 blstm2-backward.W_zr:0.946784 blstm2-backward.W_hpart:1.9615 blstm2-backward.gru_nonlin:0.870405 blstm2-backward.W_y:1.89833 blstm3-forward.W_zr:1.96695 blstm3-forward.W_hpart:2.57616 blstm3-forward.gru_nonlin:0.608609 blstm3-forward.W_y:2.66922 blstm3-backward.W_zr:1.12574 blstm3-backward.W_hpart:1.98288 blstm3-backward.gru_nonlin:0.327615 blstm3-backward.W_y:2.08487 output.affine:4.74356 output-xent.affine:6.05219 ]
LOG (nnet3-show-progress[5.2.6-e93e]:main():nnet3-show-progress.cc:150) Relative parameter differences per layer are [ blstm1-forward.W_zr:0.0389782 blstm1-forward.W_hpart:0.0329627 blstm1-forward.gru_nonlin:0.112255 blstm1-forward.W_y:0.0808113 blstm1-backward.W_zr:0.0142172 blstm1-backward.W_hpart:0.0136064 blstm1-backward.gru_nonlin:0.0742011 blstm1-backward.W_y:0.0701229 blstm2-forward.W_zr:0.0190653 blstm2-forward.W_hpart:0.0435173 blstm2-forward.gru_nonlin:0.0526738 blstm2-forward.W_y:0.0607329 blstm2-backward.W_zr:0.0185964 blstm2-backward.W_hpart:0.04245 blstm2-backward.gru_nonlin:0.0523861 blstm2-backward.W_y:0.0583726 blstm3-forward.W_zr:0.0382417 blstm3-forward.W_hpart:0.0551797 blstm3-forward.gru_nonlin:0.0371611 blstm3-forward.W_y:0.0773332 blstm3-backward.W_zr:0.0223375 blstm3-backward.W_hpart:0.0433168 blstm3-backward.gru_nonlin:0.0204331 blstm3-backward.W_y:0.0628497 output.affine:0.248664 output-xent.affine:0.113362 ]

for non-fast-bpgru:
LOG (nnet3-show-progress[5.2.105-0e5bb]:main():nnet3-show-progress.cc:139) Parameter differences per layer are [ blstm1-forward.W_z.xs_z:0.671286 blstm1-forward.W_z.xs_r:1.28809 blstm1-forward.W_h.UW:2.75 blstm1-forward.W_s.ys:2.25157 blstm1-backward.W_z.xs_z:0.503601 blstm1-backward.W_z.xs_r:0.907589 blstm1-backward.W_h.UW:2.01245 blstm1-backward.W_s.ys:2.43324 blstm2-forward.W_z.xs_z:0.976396 blstm2-forward.W_z.xs_r:0.620039 blstm2-forward.W_h.UW:2.4603 blstm2-forward.W_s.ys:2.10051 blstm2-backward.W_z.xs_z:0.785489 blstm2-backward.W_z.xs_r:0.486103 blstm2-backward.W_h.UW:2.01913 blstm2-backward.W_s.ys:1.79631 blstm3-forward.W_z.xs_z:1.90477 blstm3-forward.W_z.xs_r:0.387051 blstm3-forward.W_h.UW:2.65777 blstm3-forward.W_s.ys:2.40192 blstm3-backward.W_z.xs_z:1.17679 blstm3-backward.W_z.xs_r:0.297318 blstm3-backward.W_h.UW:2.23669 blstm3-backward.W_s.ys:2.09789 output.affine:4.98484 output-xent.affine:5.98076 ]
LOG (nnet3-show-progress[5.2.105-0e5bb]:main():nnet3-show-progress.cc:150) Relative parameter differences per layer are [ blstm1-forward.W_z.xs_z:0.0150646 blstm1-forward.W_z.xs_r:0.053782 blstm1-forward.W_h.UW:0.0584832 blstm1-forward.W_s.ys:0.0646913 blstm1-backward.W_z.xs_z:0.0113544 blstm1-backward.W_z.xs_r:0.0395913 blstm1-backward.W_h.UW:0.04334 blstm1-backward.W_s.ys:0.0703826 blstm2-forward.W_z.xs_z:0.021587 blstm2-forward.W_z.xs_r:0.0279242 blstm2-forward.W_h.UW:0.0528685 blstm2-forward.W_s.ys:0.0632478 blstm2-backward.W_z.xs_z:0.0174474 blstm2-backward.W_z.xs_r:0.022312 blstm2-backward.W_h.UW:0.0437836 blstm2-backward.W_s.ys:0.0546943 blstm3-forward.W_z.xs_z:0.0418089 blstm3-forward.W_z.xs_r:0.01732 blstm3-forward.W_h.UW:0.0577115 blstm3-forward.W_s.ys:0.0714268 blstm3-backward.W_z.xs_z:0.0260396 blstm3-backward.W_z.xs_r:0.0129631 blstm3-backward.W_h.UW:0.0491269 blstm3-backward.W_s.ys:0.0640591 output.affine:0.246963 output-xent.affine:0.105007 ]


for fast-bgru:
LOG (nnet3-show-progress[5.2.6-e93e]:main():nnet3-show-progress.cc:139) Parameter differences per layer are [ blstm1-forward.W_zr:2.09704 blstm1-forward.W_hpart:1.82114 blstm1-forward.gru_nonlin:2.81622 blstm1-backward.W_zr:1.07526 blstm1-backward.W_hpart:1.05441 blstm1-backward.gru_nonlin:2.31986 blstm2-forward.W_zr:1.00678 blstm2-forward.W_hpart:2.41631 blstm2-forward.gru_nonlin:1.20792 blstm2-backward.W_zr:0.945388 blstm2-backward.W_hpart:2.39077 blstm2-backward.gru_nonlin:1.15106 blstm3-forward.W_zr:1.60951 blstm3-forward.W_hpart:2.60522 blstm3-forward.gru_nonlin:1.14501 blstm3-backward.W_zr:1.0017 blstm3-backward.W_hpart:2.14145 blstm3-backward.gru_nonlin:0.80665 output.affine:5.95956 output-xent.affine:7.78256 ]
LOG (nnet3-show-progress[5.2.6-e93e]:main():nnet3-show-progress.cc:150) Relative parameter differences per layer are [ blstm1-forward.W_zr:0.0424694 blstm1-forward.W_hpart:0.0500089 blstm1-forward.gru_nonlin:0.0949549 blstm1-backward.W_zr:0.0219653 blstm1-backward.W_hpart:0.0300435 blstm1-backward.gru_nonlin:0.0826978 blstm2-forward.W_zr:0.0207894 blstm2-forward.W_hpart:0.0649269 blstm2-forward.gru_nonlin:0.0470626 blstm2-backward.W_zr:0.0197047 blstm2-backward.W_hpart:0.0643396 blstm2-backward.gru_nonlin:0.0449305 blstm3-forward.W_zr:0.0323507 blstm3-forward.W_hpart:0.0701869 blstm3-forward.gru_nonlin:0.0448829 blstm3-backward.W_zr:0.0207279 blstm3-backward.W_hpart:0.0593213 blstm3-backward.gru_nonlin:0.0324648 output.affine:0.220267 output-xent.affine:0.108276 ]

for non-fast-bgru:
LOG (nnet3-show-progress[5.2.105-0e5bb]:main():nnet3-show-progress.cc:139) Parameter differences per layer are [ blstm1-forward.W_z.xs_z:0.710717 blstm1-forward.W_z.xs_r:1.33378 blstm1-forward.W_h.UW:3.27705 blstm1-backward.W_z.xs_z:0.286542 blstm1-backward.W_z.xs_r:0.603063 blstm1-backward.W_h.UW:2.23348 blstm2-forward.W_z.xs_z:0.8805 blstm2-forward.W_z.xs_r:0.712553 blstm2-forward.W_h.UW:3.07543 blstm2-backward.W_z.xs_z:0.731118 blstm2-backward.W_z.xs_r:0.530496 blstm2-backward.W_h.UW:2.6328 blstm3-forward.W_z.xs_z:1.48982 blstm3-forward.W_z.xs_r:0.598289 blstm3-forward.W_h.UW:3.12194 blstm3-backward.W_z.xs_z:0.797495 blstm3-backward.W_z.xs_r:0.377583 blstm3-backward.W_h.UW:2.47269 output.affine:6.17841 output-xent.affine:7.69225 ]
LOG (nnet3-show-progress[5.2.105-0e5bb]:main():nnet3-show-progress.cc:150) Relative parameter differences per layer are [ blstm1-forward.W_z.xs_z:0.020556 blstm1-forward.W_z.xs_r:0.0375201 blstm1-forward.W_h.UW:0.080817 blstm1-backward.W_z.xs_z:0.00862262 blstm1-backward.W_z.xs_r:0.0177902 blstm1-backward.W_h.UW:0.0603691 blstm2-forward.W_z.xs_z:0.0255585 blstm2-forward.W_z.xs_r:0.02063 blstm2-forward.W_h.UW:0.0785044 blstm2-backward.W_z.xs_z:0.0210071 blstm2-backward.W_z.xs_r:0.0152342 blstm2-backward.W_h.UW:0.0685368 blstm3-forward.W_z.xs_z:0.04203 blstm3-forward.W_z.xs_r:0.0176468 blstm3-forward.W_h.UW:0.0797376 blstm3-backward.W_z.xs_z:0.0227358 blstm3-backward.W_z.xs_r:0.0108797 blstm3-backward.W_h.UW:0.0665488 output.affine:0.217359 output-xent.affine:0.102018 ]

**Comment by hlthu** [comment]
And I use ng-affine-options="maxchange=1.5" for pgru, and the result is as follows.

fast (p)gru Vs. no-fast (p)gru under TEDLIUM-v2

|       AM       | dev WER% | test WER% | #Parameters | Train TIme | Decode dev | Decode test |
| :------------: | :------: | :-------: | :---------: | :--------: | :--------: | :---------: |
|   bpgru-256    |   8.1    |    8.4    |    29.4M    |   30.7h    |    3.2h    |    5.1h     |
|    bgru-600    |   9.0    |    9.2    |    30.1M    |   28.8h    |    3.2h    |    5.2h     |
| fast-bpgru-256-1.0 |   8.4    |    8.8    |    29.4M    |   26.2h    |    2.4h    |    3.8h     |
| fast-bpgru-256-1.5 |   8.3    |    8.3    |    29.4M    |   26.3h    |    2.3h    |    4.1h     |
| fast-bgru-600-1.0  |   8.9    |    9.4    |    30.1M    |   24.0h    |    2.6h    |    4.0h     |
| fast-bgru-600-1.5  |   9.4    |    9.2    |    30.1M    |   23.7h    |    2.6h    |    4.0h     |

**Comment by GaofengCheng** [comment]
@hlthu we need more tries on 1.5 and evaluate it

**Comment by hlthu** [comment]
@GaofengCheng , I have done 1.5 with fast-gru, and the results is as below

|       AM       | dev WER% | test WER% | #Parameters | Train TIme | Decode dev | Decode test |
| :------------: | :------: | :-------: | :---------: | :--------: | :--------: | :---------: |
|   bpgru-256    |   8.1    |    8.4    |    29.4M    |   30.7h    |    3.2h    |    5.1h     |
|    bgru-600    |   9.0    |    9.2    |    30.1M    |   28.8h    |    3.2h    |    5.2h     |
| fast-bpgru-256-1.0 |   8.4    |    8.8    |    29.4M    |   26.2h    |    2.4h    |    3.8h     |
| fast-bpgru-256-1.5 |   8.3    |    8.3    |    29.4M    |   26.3h    |    2.3h    |    4.1h     |
| fast-bgru-600-1.0  |   8.9    |    9.4    |    30.1M    |   24.0h    |    2.6h    |    4.0h     |
| fast-bgru-600-1.5  |   9.4    |    9.2    |    30.1M    |   23.7h    |    2.6h    |    4.0h     |

**Comment by hlthu** [comment]
And I have test the performance of the fast-version (p)gru on swbd.


AM| train_dev | eval2000 | swbd subset | callhome subset | #P | Train_time |train_dev decode time | eval2000 decode time
-- | -- | -- | -- | --| -- | -- | -- | --
bpgru-256 | 12.20 | 13.9 | 9.4 | 18.2 | 29.9M | 124.4h | 8.6h | 7.2h
bgru-600 | 13.17 | 14.8 | 10.2 | 19.4 | 30.7M | 123.6h | 8.0h | 6.1h
fast-bpgru-256-1.0 | 12.74 | 14.3 | 9.6 | 19.0 | 29.9M | 111.4h | 6.3h | 4.9h
fast-bgru-600-1.0 | 13.36 | 14.9 | 10.1 | 19.7 | 30.7M | 99.9h | 6.0h | 5.3h
fast-bpgru-256-1.5 | 12.80 | 14.4 | 9.6 | 19.0 | 29.9M | 111.3h | 6.0h | 4.8h
fast-bgru-600-1.5 | 13.54 | 15.3 | 10.4 | 20.1 | 30.7M | 99.9h | 6.0h | 5.2h
fast-bpgru-256-s0.75 | 12.44 | 13.9 | 9.5 | 18.2 | 29.9M | 123.6h | 6.5h | 5.1h
fast-bgru-600-s0.75 | 13.62 | 15.2 | 10.5 | 19.9 | 30.7M | 116.6h | 6.9h | 5.1h

note:
* the first two lines are the (p)gru by @GaofengCheng 
* the 3-4rd lines use the combined matrix and set the max-change=1.0
* the 5-6rd lines are same as below, but  set the max-change=1.5
* the 7-8rd lines use the separated matrices, and  set the max-change=0.75

File: egs/tedlium/s5_r2/local/chain/tuning/run_blstm_1a.sh [hunk 4]
```diff
@@ -0,0 +1,240 @@
+#!/bin/bash
+
```
Comment (GaofengCheng) [inline egs/tedlium/s5_r2/local/chain/tuning/run_blstm_1a.sh:2]: @hlthu attach the results and log info?

File: egs/tedlium/s5_r2/local/chain/tuning/run_bpgru_1a.sh [hunk 5]
```diff
@@ -0,0 +1,240 @@
+#!/bin/bash
+
+# run_bpgru_1a.sh is a first attempt at an BPGRU system, based on xconfigs
+
+## how you run this
+# by default, with cleanup:
+# local/chain/tuning/run_bpgru_1a.sh
```
Comment (GaofengCheng) [inline egs/tedlium/s5_r2/local/chain/tuning/run_bpgru_1a.sh]: also here results and log info

File: egs/tedlium/s5_r2/local/chain/tuning/run_bpgru_1b.sh [hunk 6]
```diff
@@ -0,0 +1,240 @@
+#!/bin/bash
+
+# run_bpgru_1b.sh is the same as run_bpgru_1a.sh, but using the fast-pgru-layer 
+
+## how you run this
```
Comment (GaofengCheng) [inline egs/tedlium/s5_r2/local/chain/tuning/run_bpgru_1b.sh]: here results and log info

**Comment by hlthu** [comment]
@GaofengCheng I have added the results.

**Comment by LvHang** [comment]
Hi, Gaofeng

Do you think there is something wrong in line 740 of gru.py?
`
configs.append("# hpart_t")
configs.append("component-node name={0}.hpart_t component={0}.W_hpart input={1}".format(name, input_descriptor))
`
I think the hpart_t means (x_t * U^h), the input dim should be input_dim+rec_proj_dim.

Best,
Hang

**Pull request merged**